gapir/client/client.go

```diff
@@ -17,80 +17,118 @@ package client
+	// gapirAuthTokenMetaDataName is the key of the Context metadata pair that
+	// contains the authentication token. This token is common knowledge shared
+	// between GAPIR client (which is GAPIS) and GAPIR server (which is GAPIR
+	// device).
+	gapirAuthTokenMetaDataName = "gapir-auth-token"
+	// gRPCConnectTimeout is the time allowed to establish a gRPC connection.
+	gRPCConnectTimeout = time.Second * 10
+	// heartbeatInterval is the delay between heartbeat pings.
+	heartbeatInterval = time.Second * 2
 )
 
-type clientInfo struct {
-	device               bind.Device
-	arch                 device.Architecture
-	abi                  *device.ABI
-	deviceConnectionInfo deviceConnectionInfo
-	connection           gapir.Connection
-	bgConnection         *backgroundConnection
+// ReplayExecutor must be implemented by replay executors to handle some live
+// interactions with a running replay.
+type ReplayExecutor interface {
+	// HandlePostData handles the given post data message.
+	HandlePostData(context.Context, *gapir.PostData) error
+	// HandleNotification handles the given notification message.
+	HandleNotification(context.Context, *gapir.Notification) error
+	// HandleFinished is notified when the given replay is finished.
+	HandleFinished(context.Context, error) error
+	// HandleFenceReadyRequest handles when the replayer is waiting for the server
+	// to execute the registered FenceReadyRequestCallback for fence ID provided
+	// in the FenceReadyRequest.
+	HandleFenceReadyRequest(context.Context, *gapir.FenceReadyRequest) error
 }
 
-type deviceArch struct {
+// ReplayerKey is used to uniquely identify a GAPIR instance.
+type ReplayerKey struct {
 	device bind.Device
 	arch   device.Architecture
 }
 
-// ConnectionKey is used by manager to obtain a connection
-type ConnectionKey deviceArch
+// replayerInfo stores data related to a single GAPIR instance.
+type replayerInfo struct {
+	device               bind.Device
+	abi                  *device.ABI
+	deviceConnectionInfo deviceConnectionInfo
+	executor             ReplayExecutor
+	conn                 *grpc.ClientConn
+	rpcClient            replaysrv.GapirClient
+	rpcStream            replaysrv.Gapir_ReplayClient
+}
 
-// Client handles connections to GAPIR instances on devices.
-// A single Client can handle multiple connections.
+// Client handles multiple GAPIR instances identified by ReplayerKey.
 type Client struct {
-	// Mutex is needed due to the risk that reconnect may happen in another thread
-	mutex       sync.Mutex
-	clientInfos map[ConnectionKey]clientInfo
+	// mutex prevents data races when restarting replayers.
+	mutex     sync.Mutex
+	replayers map[ReplayerKey]*replayerInfo
 }
 
-// New returns a newly construct Client.
+// New returns a new Client with no replayers.
 func New(ctx context.Context) *Client {
-	client := &Client{clientInfos: map[ConnectionKey]clientInfo{}}
+	client := &Client{replayers: map[ReplayerKey]*replayerInfo{}}
 	app.AddCleanup(ctx, func() {
```

Comment:
I don't know if this should be continue to reside here. We may want to consider the app calling shutdown directly, instead of adding this to the app's cleanup list. 

If app is creating a client then it owns the client and should take care of it.

Reply:
I think `core/app` is a framework to write apps (used also by `gapit`, `apic`, etc...), and it has no idea about the app internal details. As such, I don't think we should special-case the gapir client in this framework.